### ISSUE TITLE
About directory structure scalability

### ISSUE THREAD

**Issue by mrtnbroder** [diegohaz/arc#50]
Hi there,

first of: nice work! I really like the work put into this, the overall consistency and composability is awesome!

Though, I have some concerns about the directory structure, and how it would scale when you' are faced with the reality and more and more components need to be added as either atoms/molecules etc.

Let's say I have several components that mount as content in a modal, the content-components all look and behave differently. So I'd have around 10 'molecule' components that only act as a content placeholder for one specific 'organism' (the Modal component). This doesn't really scale well when you have other 'molecules' that also live within the 'molecules' directory though. 

I hope that makes sense so far.

E.g. the folder would look like this:

```
-- atoms
---- Divider
---- ...etc.
-- molecules
---- ModalContent1
---- ModalContent2
---- ModalContent3
---- ModalContent4
---- ModalContent5
---- ModalContent6
---- ModalContent7
---- ModalContent8
---- ModalContent9
---- ModalContent10
---- Foo
---- Bar
---- Baz
---- ...many many more.
```

So my idea would be the following: organisms and molecules can have sub-directorys with components that will only be used by them. So it would look like this:

```
-- organisms
---- Modal
------ index.jsx (entry point of the modal component)
------ molecules (molecules ONLY used by Modal)
-------- ModalContent1
-------- ModalContent2
-------- ModalContent3
-------- ModalContent4
-------- ModalContent5
-------- ...etc.
```

organisms can have molecules
molecules can have atoms

This would be a scaleable solution imo, and still follow the atom design pattern in a way.

Would love to get some feedback on this. Maybe I'm misunderstanding something big!

In other words, how would one structure/handle components that are special to another component (it will/should never be used by other components).

**Comment by diegohaz** [comment]
Hi, @mrtnbroder. Thank you for the words.

The components directory is scalable for sure. Actually, Atomic Design itself is scalable. From [Pattern Lab](http://patternlab.dmolsen.com/), which this boilerplate is based on:
> Pattern Lab is designed to be flexible with regards to naming and structure. There's really no One Right Way to do this, and we think that's just fine. Name things as you please, and we'll leave the hair-splitting out of things.

For example, I've added a `themes` folder to the structure. In [this great article](https://medium.com/@yejodido/atomic-components-managing-dynamic-react-components-using-atomic-design-part-1-5f07451f261f#.aftxinyub), @yejodido introduces `ecosystems` and `environments`.

That said, you could add a `modals` folder to the structure. But I think your suggestion is better, I would name it `contents` though. I'm doing something similar on [`Icon`](https://github.com/diegohaz/arc/tree/master/src/components/atoms/Icon).

This is something worth discussing and it would be nice if other people could add their thoughts. In the end, though, you should do what better addresses your use case.

cc/ @bradfrost

**Comment by jlmitch5** [comment]
It seems to me like the `<Modal />` component should probably not be concerned with the content inside of it...it should just be concerned with things like opening/closing and the consistent interface elements that all modals share (the close button in the top right corner, the background overlay behind the modal, etc.)

This makes me think that it might it could be a good case for a [template](https://github.com/diegohaz/arc/issues/20#issuecomment-265934388)? with a definition that looks something like:

```
const Modal = ({ children }) => (
  <div style={overlay}>
    <div style={modal}>
      <div style={modaBody}>
        {children}
      </div>
    </div>
  </div>
)
```

There is a possible issue here though.  Modals are more complicated than just rendering a dom string.  They require internal state ("is the modal open or closed"), and also may need to interact with the larger environment ("open a modal if the API returns an error", "all modals need to be closed if the user's login session has timed out").

Curious what y'all think about this idea.

**Comment by diegohaz** [comment]
> It seems to me like the <Modal /> component should probably not be concerned with the content inside of it

I agree. What I did was creating a `Modal` molecule and some more specific modal organisms like `ConfirmModal` and `LoginModal`.

> "open a modal if the API returns an error", "all modals need to be closed if the user's login session has timed out

To achieve that, we need to make the modal(s) state global. I wrote a implementation of this on the redux branch inspired by [this answer](http://stackoverflow.com/questions/35623656/how-can-i-display-a-modal-dialog-in-redux-that-performs-asynchronous-actions/35641680#35641680) from @gaearon on SO.

Basically, we have a [modal store](https://github.com/diegohaz/arc/tree/cd27a82dcaa2cf0879b7de5b25840ef2d1e8d92c/src/store/modal), the [components](https://github.com/diegohaz/arc/blob/cd27a82dcaa2cf0879b7de5b25840ef2d1e8d92c/src/components/organisms/ConfirmModal/index.js) and the respective [containers](https://github.com/diegohaz/arc/blob/cd27a82dcaa2cf0879b7de5b25840ef2d1e8d92c/src/containers/ConfirmModal.js) to connect them to the store.

The usage would be something like this:
```jsx
<Button onClick={(props) => props.dispatch(modalShow('confirm'))}>
  Delete
</Button>
<ConfirmModal name="confirm" onConfirm={(props) => props.dispatch(somethingDelete())}>
  Are you sure?
</ConfirmModal>
```

An example is the [`UserButton` component](https://github.com/diegohaz/arc/blob/cd27a82dcaa2cf0879b7de5b25840ef2d1e8d92c/src/components/organisms/UserButton/index.js) component and the [`UserButton` container](https://github.com/diegohaz/arc/blob/cd27a82dcaa2cf0879b7de5b25840ef2d1e8d92c/src/containers/UserButton.js) that opens the [`LoginModal`](https://github.com/diegohaz/arc/blob/cd27a82dcaa2cf0879b7de5b25840ef2d1e8d92c/src/components/organisms/LoginModal/index.js).
